crates/core/src/rlp/decode.rs

```diff
@@ -225,63 +257,47 @@ impl<T: RLPDecode> RLPDecode for Vec<T> {
         let mut current_slice = payload;
 
         while !current_slice.is_empty() {
-            let (_, rest) = decode_bytes(current_slice)?;
-            let item = T::decode(current_slice)?;
+            let (item, rest_current_list) = T::decode_unfinished(current_slice)?;
             result.push(item);
-            current_slice = rest;
+            current_slice = rest_current_list;
         }
 
-        Ok(result)
+        Ok((result, input_rest))
     }
 }
 
 impl<T1: RLPDecode, T2: RLPDecode> RLPDecode for (T1, T2) {
-    fn decode(rlp: &[u8]) -> Result<Self, RLPDecodeError> {
+    fn decode_unfinished(rlp: &[u8]) -> Result<(Self, &[u8]), RLPDecodeError> {
         if rlp.is_empty() {
             return Err(RLPDecodeError::InvalidLength);
         }
-        let (is_list, payload, _) = decode_rlp_item(rlp)?;
+
+        let (is_list, payload, input_rest) = decode_rlp_item(rlp)?;
         if !is_list {
             return Err(RLPDecodeError::MalformedData);
         }
 
-        let (is_list, first, rest) = decode_rlp_item(payload)?;
-        let first = if first.is_empty() && is_list {
-            T1::decode(&[RLP_EMPTY_LIST])?
-        } else {
-            T1::decode(payload)?
-        };
-        let second = T2::decode(rest)?;
-        Ok((first, second))
+        let (first, first_rest) = T1::decode_unfinished(payload)?;
+        let (second, _second_rest) = T2::decode_unfinished(first_rest)?;
```

Comment:
Here we should check that `_second_rest` is empty. As an example, `0xc30101` should fail, but instead is decoded to `(1_u8, 1_u8)`